cpp/tests/io/fst/fst_test.cu

```diff
@@ -0,0 +1,265 @@
+                      R"("author": "Nigel Rees",)"
+                      R"("title": "Sayings of the Century",)"
+                      R"("price": 8.95)"
+                      R"(}  )"
+                      R"({)"
+                      R"("category": "reference",)"
+                      R"("index:" [4,{},null,{"a":[]}],)"
+                      R"("author": "Nigel Rees",)"
+                      R"("title": "Sayings of the Century",)"
+                      R"("price": 8.95)"
+                      R"(}  {} [] [ ])";
+
+  size_t string_size                 = input.size() * (1 << 10);
+  auto d_input_scalar                = cudf::make_string_scalar(input);
+  auto& d_string_scalar              = static_cast<cudf::string_scalar&>(*d_input_scalar);
+  const cudf::size_type repeat_times = string_size / input.size();
+  auto d_input_string                = cudf::strings::repeat_string(d_string_scalar, repeat_times);
+  auto& d_input = static_cast<cudf::scalar_type_t<std::string>&>(*d_input_string);
+  input         = d_input.to_string(stream);
+
+  // Prepare input & output buffers
+  constexpr std::size_t single_item = 1;
+  hostdevice_vector<SymbolT> output_gpu(input.size(), stream_view);
+  hostdevice_vector<SymbolOffsetT> output_gpu_size(single_item, stream_view);
+  hostdevice_vector<SymbolOffsetT> out_indexes_gpu(input.size(), stream_view);
+
+  // Run algorithm
+  DfaFstT parser{pda_sgs, pda_state_tt, pda_out_tt, stream.value()};
+
+  // Allocate device-side temporary storage & run algorithm
+  parser.Transduce(d_input.data(),
+                   static_cast<SymbolOffsetT>(d_input.size()),
+                   output_gpu.device_ptr(),
+                   out_indexes_gpu.device_ptr(),
+                   output_gpu_size.device_ptr(),
+                   start_state,
+                   stream.value());
+
+  // Async copy results from device to host
+  output_gpu.device_to_host(stream.view());
+  out_indexes_gpu.device_to_host(stream.view());
+  output_gpu_size.device_to_host(stream.view());
+
+  // Prepare CPU-side results for verification
+  std::string output_cpu{};
+  std::vector<SymbolOffsetT> out_index_cpu{};
+  output_cpu.reserve(input.size());
+  out_index_cpu.reserve(input.size());
+
+  // Run CPU-side algorithm
+  fst_baseline(std::begin(input),
+               std::end(input),
+               start_state,
+               pda_sgs,
+               pda_state_tt,
+               pda_out_tt,
+               std::back_inserter(output_cpu),
+               std::back_inserter(out_index_cpu));
+
+  // Make sure results have been copied back to host
+  stream.synchronize();
+
+  // Verify results
+  ASSERT_EQ(output_gpu_size[0], output_cpu.size());
+  for (std::size_t i = 0; i < output_cpu.size(); i++) {
+    EXPECT_EQ(output_gpu[i], output_cpu[i]) << "Mismatch at index #" << i;
+  }
+  for (std::size_t i = 0; i < output_cpu.size(); i++) {
+    EXPECT_EQ(out_indexes_gpu[i], out_index_cpu[i]) << "Mismatch at index #" << i;
+  }
```

Comment:
I'm not aware we'd have that. After all, it's just the for loop + streaming some extra debug info into the failing test.

Reply:
we have comparison utils for column, and tabl. No comparison utils between hostdevice_vector and vector exist.

Reply:
I guess in that case, since we have at least 4-5 places where this kind of test is added in the three PRs, I would suggest adding a generic ASSERT/EXPECT_ARRAY_EQUAL(a, b, size)

Reply:
```
// Macro for vector/array comparison
#define CUDF_TEST_EXPECT_VECTOR_EQUAL(lhs, rhs, size)          \
  do {                                                         \
    for (decltype(size) i = 0; i < size; i++)                  \
      EXPECT_EQ(lhs[i], rhs[i]) << "Mismatch at index #" << i; \
  } while (0)
```
Since this is direct macro without any function call, SCOPED_TRACE is not required.

Reply:
Thanks for the suggestion @karthikeyann and @upsj! I've adopted your macro and put it into the FST test as well as the logical stack test 👍